src/java/org/apache/cassandra/utils/Backoff.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.utils;
+
+import java.util.concurrent.TimeUnit;
+import java.util.function.DoubleSupplier;
+
+public interface Backoff
+{
+    /**
+     * @return max attempts allowed, {@code < 0} implies unbounded retries
+     */
+    int maxAttempts();
+    long computeWaitTime(int retryCount);
+    TimeUnit unit();
+
+    enum None implements Backoff
+    {
+        INSTANCE;
+
+        @Override
+        public int maxAttempts()
+        {
+            return 0;
+        }
+
+        @Override
+        public long computeWaitTime(int retryCount)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public TimeUnit unit()
+        {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    class ExpoentialBackoff implements Backoff
+    {
+        private final int maxAttempts;
+        private final long baseSleepTimeMillis;
+        private final long maxSleepMillis;
+        private final DoubleSupplier randomSource;
+
+        public ExpoentialBackoff(int maxAttempts, long baseSleepTimeMillis, long maxSleepMillis, DoubleSupplier randomSource)
+        {
+            this.maxAttempts = maxAttempts;
+            this.baseSleepTimeMillis = baseSleepTimeMillis;
+            this.maxSleepMillis = maxSleepMillis;
+            this.randomSource = randomSource;
+        }
+
+        @Override
+        public int maxAttempts()
+        {
+            return maxAttempts;
+        }
+
+        @Override
+        public long computeWaitTime(int retryCount)
+        {
+            long baseTimeMillis = baseSleepTimeMillis * (1L << retryCount);
+            // its possible that this overflows, so fall back to max;
+            if (baseTimeMillis <= 0)
+                baseTimeMillis = maxSleepMillis;
+            // now make sure this is capped to target max
+            baseTimeMillis = Math.min(baseTimeMillis, maxSleepMillis);
+
+            return (long) (baseTimeMillis * (randomSource.getAsDouble() + 0.5));
```

Comment:
nit: So the amount of random jitter is between `0.5` and `1.5`? That seems like a lot, but I also don't have much context for how other implementations of this work. In terms of the sequence though, this seems asymmetric in that we can, on the low end, halve the value after we've just multiplied by 2, but not double on the high end.

Reply:
I don't remember where I got this from... I pulled this from `Retries` but I pulled that from some place else... I think Hadoop...

> That seems like a lot
> this seems asymmetric in that we can, on the low end, halve the value after we've just multiplied by 2, but not double on the high end.

if you think about it, `baseTimeMillis` is fixed and we just double each attempt, so the range between each is `baseTimeMillis * 1 TO baseTimeMillis * 2`; so this jitter should be within that range to avoid overlap.

This is `baseTimeMillis / 2 TO baseTimeMillis * 1.5`... so this has the same effect, just offset by .5...  one advantage to this way is that the first attempt isn't exactly baseTimeMillis and maybe lower...  if we change `.5` to `1` then we get `baseTimeMillis * 1 TO baseTimeMillis * 2` and won't ever user a smaller range...

I am cool with `+ 1` rather than `0.5`, let me know your thoughts

Reply:
I'm fine w/ it as is.